In [1]:
%load_ext lab_black

In [17]:
import pandas as pd
from zipfile import ZipFile

In [3]:
# take a look at the data

file_companies_2005_Q1 = "../data/raw/company_2005_Q1.zip"
file_companies_2011_Q4 = "../data/raw/company_2011_Q4.zip"

myzip = ZipFile(file_companies_2005_Q1)
#     print(myzip.namelist())
filestream_2005 = myzip.open("company.idx")
myzip.close()

myzip = ZipFile(file_companies_2011_Q4)
filestream_2011 = myzip.open("company.idx")
myzip.close()

In [4]:
# load company list data
df_2005 = pd.read_csv(
    filestream_2005,
    sep=r" [ ]+",
    header=4,
    skiprows=[8],
    error_bad_lines=False,
    warn_bad_lines=False,
)
df_2005.info()

/Users/navy/anaconda3/envs/propulsion/lib/python3.9/site-packages/pandas/util/_decorators.py:311: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  return func(*args, **kwargs)
/Users/navy/anaconda3/envs/propulsion/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3441: FutureWarning: The warn_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)
/Users/navy/anaconda3/envs/propulsion/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3441: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 314826 entries, 0 to 314825
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Company Name  314826 non-null  object 
 1   Form Type     314825 non-null  object 
 2   CIK           314825 non-null  float64
 3   Date Filed    314825 non-null  object 
 4   File Name     314825 non-null  object 
dtypes: float64(1), object(4)
memory usage: 12.0+ MB


In [5]:
# load company list data
df_2011 = pd.read_csv(
    filestream_2011,
    sep=r" [ ]+",
    header=4,
    skiprows=[9],
    error_bad_lines=False,
    warn_bad_lines=False,
    encoding_errors="replace",
)
df_2011.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 202470 entries, 0 to 202469
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   Company Name  202470 non-null  object
 1   Form Type     202470 non-null  object
 2   CIK           202470 non-null  int64 
 3   Date Filed    202470 non-null  object
 4   File Name     202470 non-null  object
dtypes: int64(1), object(4)
memory usage: 7.7+ MB


In [6]:
# cleaning


def to_int(x):
    try:
        return int(x)
    except (pd.errors.IntCastingNaNError, ValueError):
        return -1


# 2005
df_2005["CIK"] = df_2005["CIK"].apply(lambda x: to_int(x))
df_2005["Date_filed"] = pd.to_datetime(df_2005["Date Filed"])
df_2005["Form Type"] = df_2005["Form Type"].astype("category")

# 2011
df_2011["Date_filed"] = pd.to_datetime(df_2011["Date Filed"])
df_2011["Form Type"] = df_2011["Form Type"].astype("category")

df_2005.to_csv("../data/processed/companies_2005_Q1.csv")
df_2011.to_csv("../data/processed/companies_2011_Q4.csv")

In [7]:
# select only 10-Q fillings

df_2005_10Q = df_2005[df_2005["Form Type"] == "10-Q"]
df_2011_10Q = df_2011[df_2011["Form Type"] == "10-Q"]
print(df_2005_10Q.shape, df_2011_10Q.shape)

(1522, 6) (7651, 6)


In [8]:
counts = df_2005_10Q.value_counts("Company Name")
print(sum(counts > 1) / len(counts))
counts

0.02097428958051421


Company Name
DIASENSE INC/PA                7
BCTC IV ASSIGNOR CORP          3
NORTEL NETWORKS LTD            3
BCTC V ASSIGNOR CORP           3
UNIMARK GROUP INC              3
                              ..
EVANS BOB FARMS INC            1
ETHAN ALLEN INTERIORS INC      1
ESTERLINE TECHNOLOGIES CORP    1
ESTEE LAUDER COMPANIES INC     1
streetTRACKS GOLD TRUST        1
Length: 1478, dtype: int64

In [9]:
counts = df_2011_10Q.value_counts("Company Name")
print(sum(counts > 1) / len(counts))
counts

0.01020545185980932


Company Name
Grupo International Inc                        22
THERMAL ENERGY STORAGE INC                     20
ABSOLUTE POTENTIAL, INC.                       12
WNC HOUSING TAX CREDIT FUND VI LP SERIES 6     11
WNC HOUSING TAX CREDIT FUND IV L P SERIES 2    11
                                               ..
Energy Transfer Partners, L.P.                  1
Energy Transfer Equity, L.P.                    1
Energy Telecom, Inc.                            1
Energy Recovery, Inc.                           1
vitaminspice                                    1
Length: 7447, dtype: int64

In [10]:
# What's happening with this company... ?
df_2011_10Q[df_2011_10Q["Company Name"] == "Grupo International Inc"]

Company Name Form Type      CIK  Date Filed  \
81961  Grupo International Inc      10-Q  1082576  2011-10-19   
81962  Grupo International Inc      10-Q  1082576  2011-10-19   
81963  Grupo International Inc      10-Q  1082576  2011-10-19   
81964  Grupo International Inc      10-Q  1082576  2011-10-24   
81965  Grupo International Inc      10-Q  1082576  2011-10-24   
81966  Grupo International Inc      10-Q  1082576  2011-10-24   
81967  Grupo International Inc      10-Q  1082576  2011-10-28   
81968  Grupo International Inc      10-Q  1082576  2011-10-28   
81969  Grupo International Inc      10-Q  1082576  2011-10-28   
81970  Grupo International Inc      10-Q  1082576  2011-10-28   
81971  Grupo International Inc      10-Q  1082576  2011-10-28   
81972  Grupo International Inc      10-Q  1082576  2011-10-28   
81973  Grupo International Inc      10-Q  1082576  2011-11-02   
81974  Grupo International Inc      10-Q  1082576  2011-11-02   
81975  Grupo International Inc      10-Q  1082576  2011-11-02   
81976  Grupo International Inc      10-Q  1082576  2011-11-03   
81977  Grupo International Inc      10-Q  1082576  2011-12-05   
81978  Grupo International Inc      10-Q  1082576  2011-12-05   
81979  Grupo International Inc      10-Q  1082576  2011-12-05   
81980  Grupo International Inc      10-Q  1082576  2011-12-07   
81981  Grupo International Inc      10-Q  1082576  2011-12-07   
81982  Grupo International Inc      10-Q  1082576  2011-12-07   

                                         File Name Date_filed  
81961  edgar/data/1082576/0001269678-11-000231.txt 2011-10-19  
81962  edgar/data/1082576/0001269678-11-000232.txt 2011-10-19  
81963  edgar/data/1082576/0001269678-11-000233.txt 2011-10-19  
81964  edgar/data/1082576/0001269678-11-000235.txt 2011-10-24  
81965  edgar/data/1082576/0001269678-11-000236.txt 2011-10-24  
81966  edgar/data/1082576/0001269678-11-000237.txt 2011-10-24  
81967  edgar/data/1082576/0001269678-11-000239.txt 2011-10-28  
81968  edgar/data/1082576/0001269678-11-000240.txt 2011-10-28  
81969  edgar/data/1082576/0001269678-11-000241.txt 2011-10-28  
81970  edgar/data/1082576/0001269678-11-000243.txt 2011-10-28  
81971  edgar/data/1082576/0001269678-11-000244.txt 2011-10-28  
81972  edgar/data/1082576/0001269678-11-000245.txt 2011-10-28  
81973  edgar/data/1082576/0001269678-11-000247.txt 2011-11-02  
81974  edgar/data/1082576/0001269678-11-000248.txt 2011-11-02  
81975  edgar/data/1082576/0001269678-11-000249.txt 2011-11-02  
81976  edgar/data/1082576/0001269678-11-000251.txt 2011-11-03  
81977  edgar/data/1082576/0001269678-11-000255.txt 2011-12-05  
81978  edgar/data/1082576/0001269678-11-000256.txt 2011-12-05  
81979  edgar/data/1082576/0001269678-11-000258.txt 2011-12-05  
81980  edgar/data/1082576/0001269678-11-000259.txt 2011-12-07  
81981  edgar/data/1082576/0001269678-11-000260.txt 2011-12-07  
81982  edgar/data/1082576/0001269678-11-000261.txt 2011-12-07

In [11]:
# Select only companies (a bit rough...)

is_company = ["INC", "Inc", "CORP", "LTD", "Corp", "CO", "inc"]
df_2005_10Q_inc = df_2005_10Q[
    df_2005_10Q["Company Name"].str.contains("|".join(is_company))
]
df_2011_10Q_inc = df_2011_10Q[
    df_2011_10Q["Company Name"].str.contains("|".join(is_company))
]

In [12]:
# Check the remaining entries

df_2005_10Q_fund = df_2005_10Q[
    ~df_2005_10Q["Company Name"].str.contains("|".join(is_company))
]
df_2011_10Q_fund = df_2011_10Q[
    ~df_2011_10Q["Company Name"].str.contains("|".join(is_company))
]
df_2005_10Q_fund
df_2011_10Q_fund

Company Name Form Type      CIK  \
452                                  99 CENTS ONLY STORES      10-Q  1011290   
585                          AAA CAPITAL ENERGY FUND L.P.      10-Q  1057051   
587                       AAA CAPITAL ENERGY FUND L.P. II      10-Q  1227268   
704                                   ABBOTT LABORATORIES      10-Q     1800   
1059                                  ACADIA REALTY TRUST      10-Q   899629   
...                                                   ...       ...      ...   
201422          Zhong Wen International Holding Co., Ltd.      10-Q  1494502   
202180  iShares S&P GSCI Commodity-Indexed Investing P...      10-Q  1346616   
202186           iShares S&P GSCI Commodity-Indexed Trust      10-Q  1332174   
202192                               iShares Silver Trust      10-Q  1330568   
202453                                       vitaminspice      10-Q  1382056   

        Date Filed                                    File Name Date_filed  
452     2011-11-10  edgar/data/1011290/0001140361-11-052840.txt 2011-11-10  
585     2011-11-14  edgar/data/1057051/0000950123-11-098443.txt 2011-11-14  
587     2011-11-14  edgar/data/1227268/0000950123-11-098383.txt 2011-11-14  
704     2011-11-04     edgar/data/1800/0001104659-11-061064.txt 2011-11-04  
1059    2011-11-02   edgar/data/899629/0001445305-11-003166.txt 2011-11-02  
...            ...                                          ...        ...  
201422  2011-11-14  edgar/data/1494502/0001144204-11-063700.txt 2011-11-14  
202180  2011-11-08  edgar/data/1346616/0001193125-11-302344.txt 2011-11-08  
202186  2011-11-08  edgar/data/1332174/0001193125-11-302344.txt 2011-11-08  
202192  2011-11-08  edgar/data/1330568/0001193125-11-302064.txt 2011-11-08  
202453  2011-11-21  edgar/data/1382056/0001127855-11-000628.txt 2011-11-21  

[801 rows x 6 columns]

In [13]:
df_2005_10Q_inc.drop_duplicates("Company Name", inplace=True)
df_2011_10Q_inc.drop_duplicates("Company Name", inplace=True)
print(df_2005_10Q_inc.shape, df_2011_10Q_inc.shape)

(1397, 6) (6677, 6)


/Users/navy/anaconda3/envs/propulsion/lib/python3.9/site-packages/pandas/util/_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


In [14]:
# merge to see what companies listed in 2005 are still there in 2011

df_merge = pd.merge(
    df_2005_10Q_inc,
    df_2011_10Q_inc,
    on="Company Name",
    how="inner",
    validate="one_to_one",
)
print(df_merge.shape)
print(len(df_2011_10Q_inc) - len(df_merge))

(418, 11)
6259


Out of 1397 companies listed in 2005, 418 remain in 2011.<br>
Meanwhile, 6259 (!!) companies went public after Q1 2005.

In [15]:
# save final files to CSV

df_2005_10Q_inc.to_csv("../data/processed/companies_2005_10Q.csv")
df_2011_10Q_inc.to_csv("../data/processed/companies_2011_10Q.csv")

In [20]:
# try merging on CIK instead

df_merge_2 = pd.merge(
    df_2005_10Q_inc,
    df_2011_10Q_inc,
    on="CIK",
    how="inner",
    validate="one_to_one",
)
df_merge_2.shape

(491, 11)

In [30]:
cik_counts_2005 = df_2005_10Q_inc.value_counts("CIK")
cik_counts_2011 = df_2011_10Q_inc.value_counts("CIK")

print(cik_counts_2005[cik_counts_2005 > 1])
print(cik_counts_2011[cik_counts_2011 > 1])

# no duplicate cik in either dataset

Series([], dtype: int64)
Series([], dtype: int64)


In [35]:
cik_survived = set(df_merge_2["CIK"])
len(cik_survived)

491

In [38]:
cik_died = df_2005_10Q_inc.loc[~df_2005_10Q_inc["CIK"].isin(cik_survived), "CIK"]
cik_died = set(cik_died)
len(cik_died)

906

In [50]:
companies_survived = df_2005_10Q_inc.loc[
    df_2005_10Q_inc["CIK"].isin(cik_survived), ["Company Name", "CIK"]
].reset_index(drop=True)
companies_died = df_2005_10Q_inc.loc[
    df_2005_10Q_inc["CIK"].isin(cik_died), ["Company Name", "CIK"]
].reset_index(drop=True)

In [51]:
companies_survived.head()

Company Name      CIK
0    1 800 FLOWERS COM INC  1084869
1             8X8 INC /DE/  1023731
2                 AAR CORP     1750
3  AASTROM BIOSCIENCES INC   887359
4               ABAXIS INC   881890

In [52]:
companies_died.head()

Company Name      CIK
0         1ST STATE BANCORP INC  1074078
1  3CI COMPLETE COMPLIANCE CORP   883787
2                     3COM CORP   738076
3                510152 N B LTD  1101202
4       AAC Group Holding Corp.  1311835

In [53]:
companies_survived.to_csv("../data/processed/companies_survived.csv")
companies_died.to_csv("../data/processed/companies_died.csv")